File: Java/Othello/src/othello/Logger.java
/**
 * @author Copyright (C) 2018, Christopher N. Hume. All rights reserved.
 *
 * You should have received a copy of the MIT License along with this program.
 * If not, see https://opensource.org/licenses/MIT.
 */
package othello;

public class Logger {
  private static final String sSpace = " ";

  // Level Names borrowed from log4net:
  public static final String[] levelName = {
    "OFF", "FATAL", "ERROR", "WARN", "INFO", "DEBUG", "ALL" };

  public static final byte OFF = 0;
  public static final byte FATAL = 1;
  public static final byte ERROR = 2;
  public static final byte WARN = 3;
  public static final byte INFO = 4;
  public static final byte DEBUG = 5;
  public static final byte ALL = 6;

  public static byte Level;

  Logger(byte logLevel) {
    Level = logLevel;
  }

  public static void log(String sFormat, Object... oArgs) {
    String s = String.format(sFormat, oArgs);
    System.out.print(s);
  }

  public static void logLine(String sFormat, Object... oArgs) {
    String s = String.format(sFormat, oArgs);
    System.out.println(s);
  }

  public static void logLine() {
    System.out.println();
  }

  public static void logLevel(byte level, String sFormat, Object... oArgs) {
    if (level <= Level) {
      StringBuilder sb = new StringBuilder(levelName[level]);
      sb.append(sSpace);
      sb.append(String.format(sFormat, oArgs));
      System.out.println(sb);
    }
  }

  public static void logLevelNewLine(byte level) {
    if (level <= Level)
      logLine();
  }

  public static void flushLine(StringBuilder sb) {
    if (sb.length() > 0) {
      logLine(sb.toString());
      sb.setLength(0);
    }
  }
}

File: CSharp/Square/Square/Logging/Logger/Logger.cs
//
// Copyright (C) 2010-2021, Christopher N. Hume.  All rights reserved.
//
//[2013-09-24 CNHume]Created Class
//
namespace Logging {
  using Command;

  using System;
  using static System.String;
  using System.IO;
  using System.Text;

  static class Logger {
    #region Constants
    private const String sSpace = " ";
    private const String sInfo = "info";
    private const String sLogExtensionDefault = "log";
    public const String sLogPathDefault = @"c:\ProgramData\Square\Logs";
    #endregion

    #region Enumerations
    public enum Level : byte { data, note, warn, error, failure, none };
    #endregion

    #region Fields
    private static string sLogPath;
    #endregion

    #region Properties
    public static Level LogLevel { get; set; }
    private static FileStream LogStream { get; set; }
    public static String LogPath {
      get => sLogPath;
      set {
        sLogPath = value;
        CloseLogStream();
        LogStream = OpenLogStream(sLogPath);
      }
    }
    #endregion

    #region Constructors
    static Logger() {
      LogLevel = Level.data;
      LogPath = sLogPathDefault;
    }
    #endregion

    #region Methods
    private static string combineExtension(String filename, String ext) {
      return IsNullOrEmpty(Path.GetExtension(filename)) ? $"{filename}.{ext}" : filename;
    }

    private static string combineFilename(String path, String filename, String ext) {
      var sFullFilename = combineExtension(filename, ext);
      return Path.Combine(path, sFullFilename);
    }

    private static FileStream OpenLogStream(String path) {
      var logStream = (FileStream)null;
      try {
        if (path is not null) {
          var sFullPath = combineFilename(path, Product.ProductName, sLogExtensionDefault);
          logStream = File.Open(sFullPath, FileMode.OpenOrCreate, FileAccess.Write, FileShare.Read);
          logStream.Seek(0, SeekOrigin.End);
        }
      }
      catch (Exception ex) {
        var type = ex.GetType();
      }
      return logStream;
    }

    private static void CloseLogStream() {
      if (LogStream is not null) {
        LogStream.Dispose();
        LogStream = default;
      }
    }

    public static void LogFlush() {
      if (LogStream is not null) LogStream.Flush();
    }

    public static void LogWrite(String s) {
      if (LogStream is not null && !IsNullOrEmpty(s)) {
        var encoding = new UnicodeEncoding();
        var buffer = encoding.GetBytes(s);
        LogStream.Write(buffer, 0, buffer.Length);
      }
    }

    public static void LogWriteLine() {
      LogWrite("\n");
      LogFlush();
    }

    public static void LogWriteLine(String s) {
      if (s is not null) {
        LogWrite(s);
        LogWriteLine();
      }
    }

    private static void Write(String sFormat, params Object[] oArgs) {
      var s = Format(sFormat, oArgs);
      Console.Write(s);
      LogWrite(s);
    }

    private static void WriteLine() {
      Write("\n");
      LogFlush();
    }

    private static void WriteLine(String s) {
      if (s is not null) {
        Write(s);
        WriteLine();
      }
    }

    private static void WriteLine(String sFormat, params Object[] oArgs) {
      var s = Format(sFormat, oArgs);
      WriteLine(s);
    }

    public static void Log(String sFormat, params Object[] oArgs) {
      Write(sFormat, oArgs);
    }

    public static void LogLine(String sFormat, params Object[] oArgs) {
      WriteLine(sFormat, oArgs);
    }

    public static void LogLine() {
      WriteLine();
    }

    public static void LogInfo(Level level, String sFormat, params Object[] oArgs) {
      if (LogLevel <= level) {
        var sb = new StringBuilder(sInfo);
        sb.Append(sSpace);
        sb.Append(level);
        sb.Append(sSpace);
        sb.AppendFormat(sFormat, oArgs);
        WriteLine(sb.ToString());
      }
    }

    public static void LogInfoNewLine(Level level) {
      if (LogLevel <= level)
        WriteLine();
    }
    #endregion
  }
}
